### PR TITLE
Add simplify_geometry function

### PR DESCRIPTION
**Pull request by jijichen**

### CODE REVIEW

**Comment by mbasmanova** [comment]
@jijichen Yizhe, looks like style checks are failing. Would you run 'mvn install -DskipTests -T1C' and fix any issues it reports?

https://api.travis-ci.org/v3/job/354172563/log.txt

```
[INFO] --- maven-checkstyle-plugin:2.17:check (checkstyle) @ presto-geospatial-toolkit ---
[INFO] There are 19 errors reported by Checkstyle 8.7 with checkstyle/airbase-checks.xml ruleset.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[16] (imports) AvoidStarImport: Using the '.*' form of import should be avoided - io.airlift.slice.*.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[17,8] (imports) UnusedImports: Unused import - javafx.scene.shape.Polyline.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[18] (imports) AvoidStarImport: Using the '.*' form of import should be avoided - org.locationtech.jts.geom.*.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[24] (imports) AvoidStarImport: Using the '.*' form of import should be avoided - com.facebook.presto.geospatial.GeometryUtils.*.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[113,54] (blocks) LeftCurly: '{' at column 54 should be on a new line.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[116] (regexp) RegexpMultiline: Multiple consecutive blank lines
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[137] (regexp) RegexpMultiline: Blank line before closing brace
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[144] (regexp) RegexpMultiline: Multiple consecutive blank lines
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[151,76] (blocks) LeftCurly: '{' at column 76 should be on a new line.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[187,74] (blocks) LeftCurly: '{' at column 74 should be on a new line.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[193,9] (blocks) RightCurly: '}' at column 9 should be alone on a line.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[204,82] (blocks) LeftCurly: '{' at column 82 should be on a new line.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[210,88] (blocks) LeftCurly: '{' at column 88 should be on a new line.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[230,82] (blocks) LeftCurly: '{' at column 82 should be on a new line.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[302,9] (blocks) RightCurly: '}' at column 9 should be alone on a line.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[398,77] (blocks) LeftCurly: '{' at column 77 should be on a new line.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[432,23] (misc) ArrayTypeStyle: Array brackets at illegal position.
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[455] (regexp) RegexpMultiline: Multiple consecutive blank lines
[ERROR] src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java:[462] (regexp) RegexpMultiline: Multiple consecutive blank lines
```

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 1]
```diff
@@ -181,6 +281,62 @@ private static Geometry readPolyline(SliceInput input)
...
+     * @param lineStrings A {@link LineString} or {@link MultiLineString}
+     */
+    private static void writePolyline(Geometry lineStrings,
+                                      SliceOutput output)
+    {
+        requireNonNull(lineStrings, "polyline is null");
+        requireNonNull(output, "output is null");
+
+        int partLengh = 0;
+        if (lineStrings.getGeometryType() == JTS_MULTI_LINESTRING) {
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: "Use the string.equals(Object other) function to compare strings, not the == operator.

The function checks the actual contents of the string, the == operator checks whether the references to the objects are equal. Note that string constants are usually "interned" such that two constants with the same value can actually be compared with ==, but it's better not to rely on that."

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 2]
```diff
@@ -239,12 +395,84 @@ private static Geometry readPolygon(SliceInput input)
         return GEOMETRY_FACTORY.createMultiPolygon(polygons.toArray(new Polygon[0]));
     }
 
+    private static void writePolygon(Geometry polygons, SliceOutput output) {
+        requireNonNull(polygons, "polygon is null");
+        requireNonNull(output, "output is null");
+        if (polygons.getGeometryType() != JTS_POLYGON && polygons.getGeometryType() != JTS_MULTI_POLYGON) {
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: use .equals() for strings, not ==

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 3]
```diff
@@ -239,12 +395,84 @@ private static Geometry readPolygon(SliceInput input)
...
+
+
     private static void skipEnvelope(SliceInput input)
     {
         requireNonNull(input, "input is null");
         input.skip(4 * SIZE_OF_DOUBLE);
     }
 
+
+    private static void writeEnvelope(Envelope envelope, SliceOutput sliceOutput)
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: s/sliceOutput/output

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestJtsGeometryUtils.java [hunk 4]
```diff
@@ -93,15 +93,19 @@ private static void assertGeometry(String wkt)
         Slice geometry = stGeometryFromText(utf8Slice(wkt));
         Geometry expected = WKT_READER.read(wkt);
         Geometry actual = JtsGeometryUtils.deserialize(geometry);
+        Slice serialized = JtsGeometryUtils.serialize(expected);
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestJtsGeometryUtils.java]: with this addition the flow of this method is confusing to me;

how about moving extra testing to the end of this method?

```
if (expected.isEmpty()) {
   assertTrue(deserialize(JtsGeometryUtils.serialize(expected)).isEmpty());
}
else {
   assertEquals(deserialize(JtsGeometryUtils.serialize(expected)), expected);
}
```

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestJtsGeometryUtils.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        else {
+            assertEquals(actual, expected);
+        }
+
+        // Test serialization
+        if (expected.isEmpty()) {
+            assertTrue(JtsGeometryUtils.deserialize(JtsGeometryUtils.serialize(expected)).isEmpty());
+        }
+        else {
+            assertEquals(JtsGeometryUtils.deserialize(JtsGeometryUtils.serialize(expected)), expected);
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestJtsGeometryUtils.java]: this tests that JTS-serialized geometry can be JTS-deserialized; could you also test that JTS-serialized geometry can be ESRI-deserialized?

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 6]
```diff
@@ -356,6 +387,26 @@ public static Slice stStartPoint(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+
+        if (distanceTolerance < 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "distance is negative");
+        }
+
+        if (distanceTolerance == 0) {
+            return input;
+        }
+
+        return JtsGeometryUtils.serialize(JtsGeometryUtils.simplify(input, distanceTolerance));
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: please, inline the call to `simplify`

```
JtsGeometryUtils.serialize(TopologyPreservingSimplifier.simplify(JtsGeometryUtils.deserialize(input), distanceTolerance));
```

consider, static import of TopologyPreservingSimplifier.simplify

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 7]
```diff
@@ -356,6 +387,26 @@ public static Slice stStartPoint(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
         return serialize(createFromEsriGeometry(lines.getPoint(0), reference));
     }
 
+    @Description("Returns a \"simplified\" version of the given geometry")
+    @ScalarFunction("Simplify_geometry")
+    @SqlType(GEOMETRY_TYPE_NAME)
+    public static Slice simplifygeometry(@SqlType(GEOMETRY_TYPE_NAME) Slice input,
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: s/simplifygeometry/simplifyGeometry

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 8]
```diff
@@ -356,6 +387,26 @@ public static Slice stStartPoint(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
         return serialize(createFromEsriGeometry(lines.getPoint(0), reference));
     }
 
+    @Description("Returns a \"simplified\" version of the given geometry")
+    @ScalarFunction("Simplify_geometry")
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: s/Simplify_geometry/simplify_geometry

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+            for (int i = 0; i < geometry.getNumGeometries(); i++) {
+                appendGeometry(geometry.getGeometryN(i), output);
+            }
+        } else {
+            appendGeometry(geometry, output);
+        }
+        return output.slice();
+    }
+
+    private static void appendGeometry(Geometry geometry, SliceOutput output) {
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: s/appendGeometry/writeGeometry for consistency with other methods (e.g. writePoint, writePolygon, etc.)

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+                throw new IllegalArgumentException("unsupported geometry type : " + geometry.getGeometryType());
+        }
+    }
+
+
+    /**
+     * Utilize JTS to simplifies a geometry and ensures that the result is a valid geometry having the same dimension
+     * and number of components as the input, and with the components having the same topological relationship.
+     */
+    public static Geometry simplify(Slice shape, double distanceTolerance) {
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: remove this method and move logic to GeoFunctions#simplifyGeometry

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+     *
+     * @param geometry A {@link LineString} or {@link MultiLineString}
+     */
+    private static void writePolyline(Geometry geometry,
+                                      SliceOutput output)
+    {
+        requireNonNull(geometry, "polyline is null");
+        requireNonNull(output, "output is null");
+
+        int partLengh;
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: s/partLengh/numParts for consistency with numPoints

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+
+        writeEnvelope(geometry.getEnvelopeInternal(), output);
+
+        output.writeInt(partLengh);
+
+        output.writeInt(numPoints);
+
+        if (partLengh > 0) {
+            output.writeInt(0);
+            for (int i = 1; i < partLengh; i++) {
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: nit: perhaps, loop from 0 to numParts - 1 and call getGeometryN(i) for clarity

```
for (int i = 0; i < numParts; i++) {
   output.writeInt(geometry.getGeometryN(i).getNumPoints());
}
```

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+        polygons.add(GEOMETRY_FACTORY.createPolygon(shell, holes.toArray(new LinearRing[0])));
+
+        if (polygons.size() == 1) {
+            return polygons.get(0);
+        }
+        return GEOMETRY_FACTORY.createMultiPolygon(polygons.toArray(new Polygon[0]));
+    }
+
+    private static void writePolygon(Geometry geometry, SliceOutput output) {
+        requireNonNull(geometry, "polygon is null");
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: s/polygon is null/geometry is null

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+        output.writeInt(EsriShapeType.POLYGON.code);
+
+        writeEnvelope(geometry.getEnvelopeInternal(), output);
+
+        output.writeInt(numParts);
+
+        output.writeInt(numPoints);
+
+        // Index of first point in part
+        int startIndex[] = new int[numParts + 1];
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: drop this array like you did in writePolyline and write out the lengths as you go

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 15]
```diff
@@ -181,6 +290,53 @@ private static Geometry readPolyline(SliceInput input)
         return GEOMETRY_FACTORY.createMultiLineString(lineStrings);
     }
 
+    /**
+     * serialize a geometry to a ESRI polyline.
+     * Note: both jts multilinestring and linestring corresponds to polyline.
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: s/jts/JTS

s/multilinestring/MultiLineString

s/linestring/LineString

s/corresponds/correspond

s/polyline/Polyline

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 16]
```diff
@@ -181,6 +290,53 @@ private static Geometry readPolyline(SliceInput input)
...
+     *
+     * @param geometry A {@link LineString} or {@link MultiLineString}
+     */
+    private static void writePolyline(Geometry geometry,
+            SliceOutput output)
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: move output parameter to the same line as geometry parameter

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 17]
```diff
@@ -181,6 +290,53 @@ private static Geometry readPolyline(SliceInput input)
...
+        int numPoints = geometry.getNumPoints();
+        switch (geometry.getGeometryType()) {
+            case JTS_MULTI_LINESTRING:
+                numParts = geometry.getNumGeometries();
+                break;
+            case JTS_LINESTRING:
+                numParts = numPoints > 0 ? 1 : 0;
+                break;
+            default:
+                throw new IllegalArgumentException("writePolyline can only handle multi-linestring or linestring");
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: perhaps, "Expected LineString or MultiLineString, but got " + geometry.getGeometryType()

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 18]
```diff
@@ -239,12 +395,83 @@ private static Geometry readPolygon(SliceInput input)
         return GEOMETRY_FACTORY.createMultiPolygon(polygons.toArray(new Polygon[0]));
     }
 
+    private static void writePolygon(Geometry geometry, SliceOutput output)
+    {
+        requireNonNull(geometry, "geometry is null");
+        requireNonNull(output, "output is null");
+        if (!geometry.getGeometryType().equals(JTS_POLYGON) && !geometry.getGeometryType().equals(JTS_MULTI_POLYGON)) {
+            throw new IllegalArgumentException("writePolyline can only handle multi-polygon or polygon");
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: perhaps, perhaps, "Expected Polygon or MultiPolygon, but got " + geometry.getGeometryType()

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 19]
```diff
@@ -239,12 +395,83 @@ private static Geometry readPolygon(SliceInput input)
...
+            output.writeInt(partIndex);
+            for (int i = 0; i < numGeometries; i++) {
+                Polygon polygon = (Polygon) geometry.getGeometryN(i);
+                if (polygon.getNumPoints() <= 0) {
+                    continue;
+                }
+
+                int numInteriorRing = polygon.getNumInteriorRing();
+                partIndex += polygon.getExteriorRing().getNumPoints();
+                // if it's not the last polygon or it has interior rings, write the index
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: perhaps,

```
if (i == numGeometries - 1 && numInteriorRing == 0) {
   // this is the last part
   break;
}

partIndex += polygon.getExteriorRing().getNumPoints();
output.writeInt(partIndex);
```

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 20]
```diff
@@ -239,12 +395,83 @@ private static Geometry readPolygon(SliceInput input)
...
+                if (i != numGeometries - 1 || numInteriorRing > 0) {
+                    output.writeInt(partIndex);
+                }
+
+                for (int j = 0; j < numInteriorRing; j++) {
+                    partIndex += polygon.getInteriorRingN(j).getNumPoints();
+                    // if it's not the last polygon or it's not the last interior ring, write the index
+                    if (i != numGeometries - 1 || j != numInteriorRing - 1) {
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: perhaps, 

```
if (i == numGeometries - 1 && j == numInteriorRing - 1) {
   // this is the last part
   break;
}

partIndex += polygon.getInteriorRingN(j).getNumPoints();
output.writeInt(partIndex);
```

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 21]
```diff
@@ -387,6 +388,27 @@ public static Slice stStartPoint(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
         return serialize(createFromEsriGeometry(lines.getPoint(0), reference));
     }
 
+    @Description("Returns a \"simplified\" version of the given geometry")
+    @ScalarFunction("SimplifyGeometry")
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: s/SimplifyGeometry/simplify_geometry

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 22]
```diff
@@ -387,6 +388,27 @@ public static Slice stStartPoint(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+    @SqlType(GEOMETRY_TYPE_NAME)
+    public static Slice simplifyGeometry(@SqlType(GEOMETRY_TYPE_NAME) Slice input,
+                                         @SqlType(DOUBLE) double distanceTolerance)
+    {
+        if (Double.isNaN(distanceTolerance)) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "distance is NaN");
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: s/distance/distanceTolerance

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 23]
```diff
@@ -387,6 +388,27 @@ public static Slice stStartPoint(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+        }
+
+        if (distanceTolerance < 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "distance is negative");
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: s/distance/distanceTolerance

**Comment by mbasmanova** [comment]
@jijichen Could you update commit message and PR description? E.g "Add simplify_geometry function"

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 24]
```diff
@@ -97,9 +115,52 @@ public static Geometry deserialize(Slice shape)
...
+
         return GEOMETRY_FACTORY.createGeometryCollection(geometries.toArray(new Geometry[0]));
     }
 
+    /**
+     * Serialize JTS {@link Geometry} shape into an ESRI shape
+     */
+    public static Slice serialize(Geometry geometry)
+    {
+        int spatialReferenceId = geometry.getSRID();
```
[the remark this answers is not in the document]
Comment (jijichen) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: This seems complicated

[the remark this answers is not in the document]
Comment (jijichen) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: NVM, done

**Comment by jijichen** [comment]
@mbasmanova Masha, I'm trying to add test for simplify_geometry function
however, I found the following test case will fail:
`assertFunction("ST_AsText(ST_GeometryFromText('" + "POLYGON ((2 0, 2 1, 3 1, 3 1, 2 0))" + "'))", VARCHAR, "POLYGON ((2 0, 2 1, 3 1, 3 1, 2 0))");`

Am i missing something?
`Expected :POLYGON ((2 0, 2 1, 3 1, 3 1, 2 0))
Actual   :POLYGON ((2 0, 3 1, 3 1, 2 1, 2 0))`

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 25]
```diff
@@ -146,6 +243,30 @@ private static Geometry readMultiPoint(SliceInput input)
...
     }
 
+    private static void writeMultiPoint(MultiPoint geometry, SliceOutput output, boolean isInGeometryCollection)
+    {
+        requireNonNull(geometry, "geometry is null");
+        requireNonNull(output, "output is null");
+
+        int numPoints = geometry.getNumPoints();
+        if (isInGeometryCollection) {
+            int size = SIZE_OF_INT + SIZE_OF_DOUBLE * 4 + SIZE_OF_INT + SIZE_OF_DOUBLE * 2 * numPoints;
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: `size` variable is not needed

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestJtsGeometryUtils.java [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+
+        // Test serialization
+        Slice jtsSerializedGeometry = JtsGeometryUtils.serialize(expected);
+        if (expected.isEmpty()) {
+            assertTrue(JtsGeometryUtils.deserialize(jtsSerializedGeometry).isEmpty());
+        }
+        else {
+            assertEquals(JtsGeometryUtils.deserialize(jtsSerializedGeometry), expected);
+        }
+        assertEquals(GeometryUtils.deserialize(jtsSerializedGeometry).asText(), (GeometryUtils.deserialize(geometry)).asText());
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestJtsGeometryUtils.java]: @jijichen FYI, 2.0.0 version of ESRI library added equals method to geometry collection. We are currently using 1.2.1. I'll be upgrading soon and will fix this then.

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 27]
```diff
@@ -181,6 +301,63 @@ private static Geometry readPolyline(SliceInput input)
...
+        int numPoints = geometry.getNumPoints();
+        switch (geometry.getGeometryType()) {
+            case JTS_MULTI_LINESTRING:
+                numParts = geometry.getNumGeometries();
+                break;
+            case JTS_LINESTRING:
+                numParts = numPoints > 0 ? 1 : 0;
+                break;
+            default:
+                throw new IllegalArgumentException("Expected Polygon or MultiPolygon, but got " + geometry.getGeometryType());
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: s/Polygon/LineString

s/MultiPolygon/MultiLineString

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 28]
```diff
@@ -181,6 +301,63 @@ private static Geometry readPolyline(SliceInput input)
         return GEOMETRY_FACTORY.createMultiLineString(lineStrings);
     }
 
+    /**
+     * Serialize a geometry to an ESRI polyline.
```
Comment (mbasmanova) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: This method doesn't need a comment. If you like to keep it, let's cleanup it a bit. How about this:

```
Serialize a LineString or MultiLineString to an ESRI Polyline.
```

Then, drop "Note:..." and @param.

**Comment by mbasmanova** [comment]
@jijichen Yizhe, thank you for the contribution.

**Pull request merged**